### ISSUE TITLE
BERT fine tuning

### ISSUE THREAD

**Issue by Foehnc** [leemengtaiwan/leemengtaiwan.github.io#9]
我是一名學生，感謝您關於bert的分享。我跟著博客做了一遍，但是不知道為什麼模型總是把結果預測為最後一類，儘管損失一直在下降，並且當訓練很多epoch之後，模型對於任何樣例在所有類上的預測概率相同，請問您有出現過這種問題嗎？或者您知道這個問題發生可能的原因嗎？

**Comment by leemengtw** [comment]
@ShallTearchen 我找到問題發生的原因了。先說結論，你現在再去跑一次我部落格上更新過的程式碼應該就沒有同樣問題了：

- [進擊的 BERT：NLP 界的巨人之力與遷移學習](https://leemeng.tw/attack_on_bert_transfer_learning_in_nlp.html)

幾個月前我在撰寫 BERT 文章時，HuggingFace 團隊的 [transformers](https://github.com/huggingface/transformers) 的 `model` 裡頭的參數順序分別為：

- `input_ids` (對應到 `tokens_tensors`)
- `token_type_ids` (對應到 `segments_tensors`)
- `attention_mask` (對應到 `masks_tensors`)                       

因此當初我在建立 dataset 時也依此順序回傳這些 tensors，並在 `get_prediction` 函式裡頭用以下的方式將這些 tensors 依照順序傳入，計算 accuracy：

```python
for data in dataloader:
	...
	outputs = model(*data[:3])
	...
```

但最近一次改版 `model` 實作的參數順序被改變，2 跟 3 顛倒：
- `input_ids` (對應到 `tokens_tensors`)
- `attention_mask` (對應到 `masks_tensors`)  
- `token_type_ids` (對應到 `segments_tensors`)

導致 `get_prediction` 計算 accuracy 的時候出錯，但又因為在 train loop 裡頭我是用 keyword args 的方式指定 tensors，沒有受到這變動的影響：

```python
...

outputs = model(input_ids=tokens_tensors, 
                token_type_ids=segments_tensors, 
                attention_mask=masks_tensors, 
                labels=labels)

loss = outputs[0]
# backward
loss.backward()
...
```

這導致模型實際有在被訓練（loss 下降），但 accuracy 看似沒有下降的詭異情境（畢竟計算 accuracy 時傳入的 tensor 順序有誤）。所以我將 `get_prediction` 裡頭的程式碼也使用 keyword args 的形式實作以後，自然就能正確計算 accuracy，accuracy 也能輕易突破 95%：

```python
for data in dataloader:
	...
	tokens_tensors, segments_tensors, masks_tensors = data[:3]
    outputs = model(input_ids=tokens_tensors, 
                    token_type_ids=segments_tensors, 
                    attention_mask=masks_tensors)
	...
```

以上供你參考。Happy BERTing!

**Comment by Foehnc** [comment]
@leemengtaiwan 非常感谢您在百忙之中解答我的问题！期待您之后的博客！
